Support using custom NuGet repo and basic auth for it

zhongyi-zhang opened the pull request:
Solving #361.
This PR is to support using custom NuGet repo. And for private NuGet repo, we can use a netrc file which is the bazel supported auth style for a repository rule: https://bazel.build/rules/lib/repo/utils#read_netrc.

Some more notes:

* This PR only implements single NuGet repo for a `nuget_repo` target, which is a little different from the Paket usage which can set multiple resources (as repos). But I think that's acceptable as we know which repo to get the packages in most cases.

* Only basic auth is implemented in this PR. We can expose the auth type and implement other types of auth later.

* The `resource` parameter is for a package url template. It is normal for v2 repo but a little strange for v3 repo. There is an option: we can distinguish v2 or v3 repo. For v3 repo, we fetch the basic package url from the index. Not sure if it's worth.

If the change looks good to you, I am happy to update the responding documents as well.

purkhusid commented:
I'll take a look next week, on vacation now.

purkhusid commented:
I have not forgotten about this. Will take a look this week.

purkhusid commented:
@zhongyi-zhang I did some changes to your approach and added them to this branch. Could you try this out with your private registry and see if everything works?

zhongyi-zhang commented on dotnet/private/rules/nuget/nuget_archive.bzl:
```diff
@@ -276,15 +282,64 @@ def _process_key_and_file(groups, key, file):
+    # If there is an netrc file declared for the specific package
+    # we use that one instead of the user netrc
+    if ctx.attr.netrc:
+        netrc = read_netrc(ctx, netrc)
```
It should be `netrc = read_netrc(ctx, ctx.attr.netrc)`, right?

zhongyi-zhang commented on tools/paket2bazel/Gen.fs:
```diff
@@ -40,8 +40,17 @@ let generateTarget (group: Group) (repoName: string) =
+
         sb.Append(
-            $"{i}        (\"{package.name}\", \"{package.version}\", \"{package.sha512sri}\", [{packageDeps}], [{overrides}]),\n"
+            $"{i}        (\"{package.name}\", \"{package.version}\", \"{package.sha512sri}\", [{sources}], None, [{packageDeps}], [{overrides}]),\n"
```
How about exposing a flag to set the specific netrc file path? It could be more friendly than generating `None` to save an extra step updating the generated packages.

purkhusid replied:
Added a `--netrc-file-label` to the paket2bazel tool. You can now run it with `--netrc-file-label //some:label` and the netrc file will be added to the generated nuget_repo.

zhongyi-zhang commented:
@purkhusid thank you for improving this PR a lot! Appreciate that you also updated the paket2bazel tool. I tried it with my private registry. Yes, it works. And added some comments.

purkhusid commented:
Could you try it again with the flag I added and see how you like it? If you feel like this is a decent solution to this problem I'll go ahead and merge.

zhongyi-zhang commented:
@purkhusid it all looks good to me. Thank you! Please go ahead merging it.

zhongyi-zhang commented on dotnet/private/rules/nuget/nuget_archive.bzl:
```diff
@@ -383,7 +383,7 @@ nuget_archive = repository_rule(
     attrs = {
         "sources": attr.string_list(),
-        "netrc": attr.string(),
+        "netrc": attr.label(),
```
Oh, wait, this change seems to be breaking absolute netrc path usage. I got err like this: `//external:nuget.system.text.json.v7.0.3: invalid label '/tmp/.netrc' in attribute 'netrc' in 'nuget_archive' rule: invalid target name '/tmp/.netrc': target names may not start with '/'`.

zhongyi-zhang replied:
And relative path does not seem to be working, because the default base dir is the cache dir for external pkg:
`ERROR: <my-bazel-cache-dir-for-external-pkg>/microsoft.netcore.app.runtime.linux-x64/7.0.8/BUILD.bazel:5:15: @mdmclient_deps//microsoft.netcore.app.runtime.linux-x64/7.0.8:7.0.8 depends on @nuget.microsoft.netcore.app.runtime.linux-x64.v7.0.8//:data in repository @nuget.microsoft.netcore.app.runtime.linux-x64.v7.0.8 which failed to fetch. no such package '@nuget.microsoft.netcore.app.runtime.linux-x64.v7.0.8//': Not a regular file: <my-bazel-cache-dir-for-external-pkg>/.netrc`

zhongyi-zhang replied:
It works with relative path by putting the netrc file under <WORKSPACE>/external/.

purkhusid replied:
Yes, this is a label to a file within the workspace. So e.g. if you have a `<WORKSPACE>/.netrc` you should put `//:netrc` into this attribute. I wasn't sure if I wanted to allow paths because relative paths won't work since the working directory is always under the output root and thus only absolute paths would work.

purkhusid replied:
Doesn´t the user .netrc fill the need for a global netrc? We default to it if there is not netrc declared in the attribute.

zhongyi-zhang replied:
The `read_user_netrc` method reads `${HOME}/.netrc`. In certain pipeline build scenario, the build tasks run in build agents. Bringing in user-relevant var like `${HOME}` is uncommon.
But I agree that using label to a file is more bazel nature comparing to a string. Probably I should just override `${HOME}` in my scenario.
